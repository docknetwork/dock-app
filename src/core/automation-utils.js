export function addTestId(testId) {
  return {
    testID: testId,
    accessibilityLabel: testId,
  };
}
