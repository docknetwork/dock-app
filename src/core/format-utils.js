export function formatCurrency(value, currency = 'USD') {
  var formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,

    minimumFractionDigits: 2,
    maximumFractionDigits: 10,
  });

  return formatter.format(value);
}

export function formatAddress(value, size = 19) {
  if (!value || size > value.length) {
    return value;
  }

  const offset = size / 2;
  return `${value.substring(0, offset)}...${value.substring(
    value.length - offset,
  )}`;
}
