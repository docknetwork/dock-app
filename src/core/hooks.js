import {useEffect} from 'react';
import {BackHandler} from 'react-native';

export function useBackHandler(callback) {
  useEffect(() => {
    const backHandler = BackHandler.addEventListener(
      'hardwareBackPress',
      () => {
        console.log('stuff');

        return true;
      },
    );

    return () => backHandler.remove();
  }, []);
}
