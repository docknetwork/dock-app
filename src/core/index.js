/**
 * JSON.stringify with error handler
 *
 * @param value
 * @returns {string}
 */
export function getJsonOrError(value) {
  try {
    return JSON.stringify(value);
  } catch (err) {
    return err;
  }
}
