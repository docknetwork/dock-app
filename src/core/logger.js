const winston = require('winston');

export const Logger = winston.createLogger({
  format: winston.format.simple(),
  levels: ['debug'],
  transports: [new winston.transports.Console()],
});
