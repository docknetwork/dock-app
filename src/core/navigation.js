import {createNavigationContainerRef} from '@react-navigation/native';

let history = [];

export const navigationRef = createNavigationContainerRef();

export const getNavigationHistory = () => history;
export const clearNavigationHistory = () => (history = []);

export function navigate(name, params, skipFromHistory) {
  if (navigationRef.isReady()) {
    navigationRef.navigate(name, params);
    if (skipFromHistory) {
      return;
    }
    history.push({
      name,
      params,
    });
  }
}

export function getHistory() {
  return history;
}

export function navigateBack() {
  history.pop();

  if (!history.length) {
    return;
  }

  const {name, params} = history.pop();
  navigationRef.goBack();
  // navigate(name, params);
}
