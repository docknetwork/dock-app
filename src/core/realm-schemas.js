export const TokenPrice = {
  name: 'TokenPrice',
  properties: {
    symbol: 'string',
    name: 'string?',
    price: 'float',
  },
  primaryKey: 'symbol',
};

export const Transaction = {
  name: 'Transaction',
  properties: {
    id: 'string',
    hash: 'string?',
    date: 'date',
    fromAddress: 'string',
    recipientAddress: 'string',
    amount: 'string',
    feeAmount: 'string',
    status: 'string',
  },
  primaryKey: 'id',
};

export const Account = {
  name: 'Account',
  properties: {
    id: 'string',
    name: 'string',
    balance: 'string?',
  },
  primaryKey: 'id',
};
