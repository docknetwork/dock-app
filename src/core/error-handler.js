import { Stack } from 'native-base';
import React from 'react';
import { Typography } from 'src/design-system';
import { translate } from 'src/locales';

export class ErrorBoundary extends React.Component<any, any> {
   constructor(props: any) {
      super(props);
      this.state = { hasError: false };
   }

   static getDerivedStateFromError(error: any) {
      return { hasError: true };
   }

   componentDidCatch(error: any, errorInfo: any) {
      // TODO: log the error in sentry
      console.error(error);
   }

   render() {
      if (this.state.hasError) {
         return <Stack><Typography>{translate('global.rendering_error')}</Typography></Stack>;
      }

      return this.props.children;
   }
}

export const withErrorBoundary = (WrapperComponent: any) => (props: any) => {
   return (
      <ErrorBoundary>
         <WrapperComponent {...props} />
      </ErrorBoundary>
   );
};
