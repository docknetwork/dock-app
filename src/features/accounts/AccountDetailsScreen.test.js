import {shallow} from 'enzyme';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import {
  AccountDetailsScreen,
  filterTransactionHistory,
} from './AccountDetailsScreen';
import {
  transactionsOperations,
  TransactionStatus,
} from '../transactions/transactions-slice';
import thunk from 'redux-thunk';
import Realm from 'realm';
import {Transaction} from '../../core/realm-schemas';

const mockStore = configureMockStore([thunk]);

const initMockRealm = () => {
  return Realm.open({
    path: 'dock_unit_test',
    schema: [Transaction],
    schemaVersion: 3,
    deleteRealmIfMigrationNeeded: false,
    inMemory: true,
    migration: () => {
      // No migration required so far
    },
  });
};

const initMockTransactions = () => {
  const today = new Date();
  const yesterday = new Date(today);
  yesterday.setDate(yesterday.getDate() - 1);
  return [
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '0',
      hash: null,
      network: 'testnet',
      status: TransactionStatus.Complete,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '1',
      hash: null,
      network: 'testnet',
      status: TransactionStatus.InProgress,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '2',
      hash: null,
      network: 'testnet',
      status: TransactionStatus.Failed,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '3',
      hash: '',
      network: 'testnet',
      status: TransactionStatus.Complete,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '4',
      hash: '',
      network: 'testnet',
      status: TransactionStatus.InProgress,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '5',
      hash: '',
      network: 'testnet',
      status: TransactionStatus.Failed,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },

    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '6',
      hash: '0xa3b3bf9d13dd726c1e0051d48cb99fd05e79442b2cda05374e898351c3ade9c2',
      network: 'testnet',
      status: TransactionStatus.Complete,
      date: new Date('2022-03-01T17:52:03.741Z'),
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '7',
      hash: '0xc3b3bf9d13dd726c1e0051d48cb99fd05e79442b2cda05374e898351c3ade9c2',
      network: 'testnet',
      status: TransactionStatus.InProgress,
      date: yesterday,
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB',
      id: '8',
      hash: '0xb3b3bf9d13dd726c1e0051d48cb99fd05e79442b2cda05374e898351c3ade9c2',
      network: 'testnet',
      status: TransactionStatus.Failed,
      date: today,
    },
    {
      amount: '10',
      feeAmount: '1',
      recipientAddress: '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsBWrong',
      fromAddress: '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsBWrong',
      id: '9',
      hash: '0xb3b3bf9d13dd726c1e0051d48cb99fd05e79442b2cda05374e898351c3ade9c2',
      network: 'testnet',
      status: TransactionStatus.Failed,
      date: today,
    },
  ];
};

describe('AccountDetailsScreen', () => {
  it('should render correctly', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(
      <AccountDetailsScreen
        account={{
          meta: {
            name: 'Test account',
            balance: 0,
          },
        }}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  describe('expect to filter transaction history items', () => {
    const address1 = '3C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB';
    const address2 = '4C7Hq5jQGxeYzL7LnVASn48tEfr6D7yKtNYSuXcgioQoWWsB';
    const txSent = {
      amount: '1',
      fromAddress: address1,
      recipientAddress: address2,
      status: 'complete',
    };
    const txReceived = {
      amount: '1',
      fromAddress: address2,
      recipientAddress: address1,
      status: 'complete',
    };
    const txReceivedFailed = {
      amount: '1',
      fromAddress: address2,
      recipientAddress: address1,
      status: 'failed',
    };

    it('expect to hide received transactions with failed status', () => {
      const transactions = filterTransactionHistory(
        [txReceived, txReceivedFailed],
        address1,
      );
      expect(transactions.length).toBe(1);
    });

    it('expect to add sent flag', () => {
      const transactions = filterTransactionHistory(
        [txReceived, txSent],
        address1,
      );
      expect(transactions[0].sent).toBe(false);
      expect(transactions[1].sent).toBe(true);
    });
  });
});
