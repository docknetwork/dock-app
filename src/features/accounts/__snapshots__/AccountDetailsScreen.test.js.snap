// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDetailsScreen should render correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#18181B",
      "flex": 1,
    }
  }
  testID="AccountDetailsScreen"
>
  <AppGlobalHeader />
  <Styled(View)>
    <Box
      alignItems="center"
      flexDirection="row"
      marginLeft={1}
      marginRight={22}
    >
      <Memo(ForwardRef(Box))
        width="80px"
      >
        <BackButton
          onPress={[Function]}
        />
      </Memo(ForwardRef(Box))>
      <Memo(ForwardRef(Box))
        alignContent="center"
        alignItems="center"
        flex={1}
        pl={15}
        width="100%"
      >
        <Styled(Component)
          variant="h3"
        />
      </Memo(ForwardRef(Box))>
      <Memo(ForwardRef(Box))
        alignItems="flex-end"
        width="80px"
      >
        <Memo(ForwardRef(Pressable))
          _pressed={
            Object {
              "opacity": 0.2,
            }
          }
          onPress={[Function]}
        >
          <SvgDotsVertical
            height="22px"
            width="22px"
          />
        </Memo(ForwardRef(Pressable))>
      </Memo(ForwardRef(Box))>
    </Box>
  </Styled(View)>
  <ForwardRef
    refreshControl={<RefreshControlMock />}
  >
    <Memo(ForwardRef(Stack))
      flex={1}
      mx={26}
    >
      <Memo(ForwardRef(Stack))
        alignItems="center"
        backgroundColor="#27272A"
        borderRadius={8}
        direction="column"
        p="32px"
      >
        <Memo()
          size={48}
        />
        <Component>
          <Component />
        </Component>
        <Memo(ForwardRef(Stack))
          direction="row"
          mt={5}
          width="100%"
        >
          <Memo(ForwardRef(Button))
            flex={1}
            onPress={[Function]}
            size="sm"
          >
            Send
          </Memo(ForwardRef(Button))>
          <Memo(ForwardRef(Button))
            flex={1}
            ml={2}
            onPress={[Function]}
            size="sm"
          >
            Receive
          </Memo(ForwardRef(Button))>
        </Memo(ForwardRef(Stack))>
      </Memo(ForwardRef(Stack))>
      <Memo(ForwardRef(Stack))
        mt={8}
      >
        <Memo(ForwardRef(Box))
          borderBottomColor="#3F3F46"
          borderBottomWidth={0.5}
          pb={4}
        >
          <Styled(Component)
            variant="h2"
          >
            Transactions
          </Styled(Component)>
        </Memo(ForwardRef(Box))>
        <Memo(ForwardRef(Box))
          mt={3}
        >
          <TransactionHistory />
        </Memo(ForwardRef(Box))>
      </Memo(ForwardRef(Stack))>
      <Memo(ForwardRef(Stack))
        backgroundColor="rgba(120, 53, 15, 0.3)"
        mt={20}
        p="16px"
      >
        <Memo(ForwardRef(Stack))
          direction="row"
        >
          <Memo(ForwardRef(Box))
            mr={3}
            mt="3px"
          >
            <SvgAlert />
          </Memo(ForwardRef(Box))>
          <Styled(Component)
            fontSize={17}
            ml={2}
            variant="h3"
          >
            Account not backed up
          </Styled(Component)>
        </Memo(ForwardRef(Stack))>
        <Memo(ForwardRef(Box))
          mt={2}
        >
          <Styled(Component)
            color="rgba(254, 243, 199, 1)"
          >
            This phrase allows you to recover your account if your phone is lost or you change your device.
          </Styled(Component)>
        </Memo(ForwardRef(Box))>
        <Memo(ForwardRef(Button))
          alignSelf="flex-start"
          backgroundColor="rgba(120, 53, 15, 1)"
          mt={4}
          size="sm"
        >
          Back up now
        </Memo(ForwardRef(Button))>
      </Memo(ForwardRef(Stack))>
    </Memo(ForwardRef(Stack))>
  </ForwardRef>
  <AccountSettingsModal
    onClose={[Function]}
  />
  <QRCodeModal
    description="Scan this QR code in your new device"
    onClose={[Function]}
    title="Export account"
  />
</RNCSafeAreaView>
`;
